### ISSUE TITLE
force_stop() seems unusuable from inside the objective function

### ISSUE THREAD

**Issue by apendleton** [adwhit/rust-nlopt#11]
Per the docs, it looks like the right way to handle an error that occurs within an objective function during `.optimize()` is to call `.force_stop()`:

> In certain cases, the caller may wish to force the optimization to halt, for some reason unknown to NLopt. For example, if the user presses Ctrl-C, or there is an error of some sort in the objective function. In this case, it is possible to tell NLopt to halt the optimization gracefully, returning the best point found so far, by calling this function from within your objective or constraint functions.

However, currently `.force_stop()` requires a mutable reference to the optimizer, and if one of those exists (like, in a closure or something), it's impossible to call `.optimize()`, so it doesn't seem like this pattern is actually possible.

More generally, even if it were, it's pretty awkward, because at the time the closure or whatnot that's being used as the objective function is created, the optimizer doesn't exist yet, so there's not really a clean way to build this functionality. I'm not sure what the best way is to address that, though maybe the most idiomatic would be some way to have an objective function with the signature `(&[f64], Option<&mut [f64]>, T) -> Result<f64, E>` that automatically does the `force_stop` behind the scenes on error?

**Comment by adwhit** [comment]
Fair point, I don't really see how `force_stop(&mut self)` can be used without breaking Rust's aliasing guarantees. I like your suggestion of returning a Result type from the objective fn, however maybe we could use a `[Termination](https://doc.rust-lang.org/stable/std/process/trait.Termination.html)`-style trait so that either `-> f64` or `-> Result<f64, E>` would work

**Comment by apendleton** [comment]
Yeah, I like that idea. I think ideally if it force-stopped because the objective function errored, you would get the error back as part of the result of optimize, so maybe the `ForceStop` variant of `FailState` becomes `ForceStop(Option<E>)` or something, and can return the error if there was any, if it's possible to wrangle the generics and traits to make that doable (I think so?). Alternatively we could have a separate variant `ObjectiveError(E)` for the case where it's a force stop and there was an error, if that feels better.
